### DISCUSSION TITLE
FYI, it looks like your cmake script does not allow for easy building of hidapi with libusb.

### DISCUSSION THREAD

**Discussion by Youw** [libusb/hidapi#682] [Q&A]
_Originally posted by @madebr in https://github.com/libusb/hidapi/issues/681#issuecomment-2222613920_:

For this, SDL uses a [`FindLibUSB.cmake` module](https://github.com/libsdl-org/SDL/blob/main/cmake/FindLibUSB.cmake) and [a GitHub prebuilt release](https://github.com/libsdl-org/SDL/commit/628e083e02703c4bbec5de909d17a874fe691880) to provide binaries

**Comment by Youw** [comment]
> FYI, it looks like your cmake script does not allow for easy building of hidapi with libusb.

Define what do you expect to be easy.
I'm using [libusb-cmake](https://github.com/libusb/libusb-cmake) and because of [this](https://github.com/libusb/hidapi/blob/4578ea24613af03e5631e400ed1167462702a112/libusb/CMakeLists.txt#L11-L12) (need `add_subdirectory(libusb)` before `add_subdirectory(hidapi)` ) - it is quite trivial.

**Comment by Youw** [comment]
_Originally posted by @madebr in https://github.com/libusb/hidapi/issues/681#issuecomment-2222978679_
            

> Define what do you expect to be easy. 

By easy, I mean the cmake script does need no modifications and can build with libusb support (using a system library) without needing to building libusb yourself.
https://github.com/libusb/hidapi/blob/4578ea24613af03e5631e400ed1167462702a112/CMakeLists.txt#L40-L48
https://github.com/libusb/hidapi/blob/4578ea24613af03e5631e400ed1167462702a112/src/CMakeLists.txt#L110-L130
`HIDAPI_WITH_LIBUSB` is not used on Windows

> I'm using [libusb-cmake](https://github.com/libusb/libusb-cmake) and because of [this](https://github.com/libusb/hidapi/blob/4578ea24613af03e5631e400ed1167462702a112/libusb/CMakeLists.txt#L11-L12) (need `add_subdirectory(libusb)` before `add_subdirectory(hidapi)` ) - it is quite trivial.

It works, but you're effectively using a fork and do not use upstream.
For SDL's use case, we don't want this hard dependency: users should bring their own libusb.
This is also valid for every (mingw) distribution (e.g. msys2)


> One more thing - do you thing we could update HIDAPI CI script to catch this/similar issues in the future?

Looking at SDL's cmake script, we've added a `SDL_WERROR` cmake option to build with `-Werror`/`/WX` (default off, enabled on ci).
Recent CMake versions have a [cmake property for this](https://cmake.org/cmake/help/latest/prop_tgt/COMPILE_WARNING_AS_ERROR.html): configure with `-DCMAKE_COMPILER_WARNING_AS_ERROR=1`.
We don't pass a special MSVC warning flag.

**Comment by Youw** [comment]
> By easy, I mean the cmake script does need no modifications and can build with libusb support (using a system library) without needing to building libusb yourself.

> HIDAPI_WITH_LIBUSB is not used on Windows

> This is also valid for every (mingw) distribution (e.g. msys2)

libusb backend for HIDAPI officially supported only on a number of platforms but Windows is not one of them. If you happen to use libusb backend of HIDAPI on Windows - you must be using modified or your own build script for HIDAPI, which is not official distribution as of yet.

On those supported platforms, libusb officially distributed in a way that pkg-config should be used to find it, and [this](https://github.com/libusb/hidapi/blob/4578ea24613af03e5631e400ed1167462702a112/libusb/CMakeLists.txt#L14-L16) part of HIDAPI CMake script handles it, at least according to my testing. Are you saying it doesn't work in your environemnt?


> Looking at SDL's cmake script, we've added a SDL_WERROR cmake option to build with -Werror//WX (default off, enabled on ci).

We have [`/WX`](https://github.com/libusb/hidapi/blob/4578ea24613af03e5631e400ed1167462702a112/.github/workflows/builds.yml#L8) on our CI too, so I wounder why it wasn't caught...

**Reply by madebr** [reply]
> On those supported platforms, libusb officially distributed in a way that pkg-config should be used to find it, and [this](https://github.com/libusb/hidapi/blob/4578ea24613af03e5631e400ed1167462702a112/libusb/CMakeLists.txt#L14-L16) part of HIDAPI CMake script handles it, at least according to my testing. Are you saying it doesn't work in your environemnt?

pkg-config is a unix-ism, and is not commonly installed on MSVC development systems.
The "CMake way" is to use modules.
(SDL's `FindLibUSB.cmake` [uses pkg-config as a hint](https://github.com/libsdl-org/SDL/blob/875c4f0a4ce2003ee64db6218526664fb718d618/cmake/FindLibUSB.cmake#L6-L10))
I know there's pkgconf, which works well on Windows, but building with MSVC in autoconf in a bash shell is an adventure on itself.
But again, the ["official" libusb MSVC binaries](https://github.com/libusb/libusb/releases/download/v1.0.27/libusb-1.0.27.7z) don't contain pc files so pkgconf won't find anything.

**Reply by Youw** [reply]
Again, libusb backend is not supported by HIDAPI on Windows, specially with MSVC. What is your use-case?

**Reply by madebr** [reply]
Oh, SDL3 allows me to configure and build in this configuration.
I assumed this was a working combination.

**Reply by Youw** [reply]
Maybe it is working in some environments, but it is not officially supported by HIDAPI build system or sources (yet).

**Reply by madebr** [reply]
I think SDL heavily patched hidapi. That's why there lives a fork inside SDL.

**Comment by madebr** [comment]
I caught this on a local MSVC instance, which apparently warned about this.
I don't know how to reproduce the warning in a fresh project (apart from adding `/W3`).
On gcc, you can warn with `-Wconversion` and `-Wfloat-conversion`.
